Userland/Libraries/LibWeb/HTML/HTMLDialogElement.cpp

```diff
@@ -93,10 +107,30 @@ WebIDL::ExceptionOr<void> HTMLDialogElement::show_modal()
     if (!document().top_layer_elements().contains(*this))
         document().add_an_element_to_the_top_layer(*this);
 
-    // FIXME: 9. Set this's close watcher to the result of establishing a close watcher given this's relevant global object, with:
-    //            - cancelAction being to return the result of firing an event named cancel at this, with the cancelable
-    //              attribute initialized to true.
-    //            - closeAction being to close the dialog given this and null.
+    // 9. Set this's close watcher to the result of establishing a close watcher given this's relevant global object
+    m_close_watcher = CloseWatcher::establish(*document().window());
+    // - cancelAction given canPreventClose being to return the result of firing an event named cancel at this, with the cancelable attribute initialized to canPreventClose.
+    auto cancel_callback_function = JS::NativeFunction::create(
+        realm(), [this](JS::VM& vm) {
+            auto event_value = vm.argument(0);
+            bool can_prevent_close = MUST(event_value.get(vm, "cancelable")).as_bool();
+            dispatch_event(DOM::Event::create(realm(), HTML::EventNames::cancel, { .cancelable = can_prevent_close }));
+            // FIXME: We should call prevent_default() on the close watcher cancel event if the dialog cancel event was.
```

Comment:
I tried, but it would crash. Ideally I'd be able to cast to an event object and use that directly, but I also couldn't get that working?

```c++
bool should_continue = dispatch_event(DOM::Event::create(realm(), HTML::EventNames::cancel, { .cancelable = can_prevent_close }));
if (!should_continue)
    MUST(MUST(event_value.get_method(vm, "prevent_default"))->internal_call(event_value, {}));
```

Happy to leave this bit to a follow up.